### PR TITLE
Implemented signing in frontend and backend, updated dockerfile for transactions

### LINKED ISSUE #86: Implement signing in transaction
**Description:**

Implement a signing system in the front end and back end,

The front end generates RSA public key and private key pairs on user creation and signs the transaction id using the private key whenever a new transaction is created by the user. The public key is sent to the back end on user creation to be stored in the user database.

The back end is responsible for verifying transaction signatures using the sending user's public key (publically available through the user service). This will ensure that the transaction came from the correct sender and was not altered after creation.

**Technical task breakdown:**

Front end: 
- Create RSA key-pair on user creation. store the private key in the front end client and send the public key to the user service so that it can be stored and retrieved from the user database
- Implement ID signing functionality whenever a transaction is created by the sender.

Back end:
- Create verifying functionality where the transaction service verifies the transaction signature using the sender's public key before sending it off for mining.

**Acceptance Criteria:**

- [x] Front end able to generate RSA key pair
- [x] Front end signs the transaction id properly
- [x] Back end can verify the signature successfully
- [x] Appropriate tests for all of the above

### PR DESCRIPTION
**Pull request by akshaysharma21**
how to test:
build and run containers as per usual to test.

sample payload to test with postman:

`{
    "from": "-----BEGIN PUBLIC KEY-----\nMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDnO7/dQaeEsLKa4KN703PR1JDx\nBbFAi10+ofEo0PeFUKgRdDt5tBUx/XrfCpiZmLTNQcRDBtNXadtp5TaBgeyb0N2w\n2TwCElB51bpJhn/haTwkXLipxnYdZkYtdvJ03SxYZ9HBUXhnCcTVfgFDHabLhocd\nYJ6CNECveFJr0mYm4wIDAQAB\n-----END PUBLIC KEY-----",
    "to" : "687",
    "amount": 10,
    "id" : "d94877994bf39b0bd2cbec2547d13e8650a5141b3e070ef7d7169ef89bc9732e",
    "signature" : "d72eb3a1afcd9f06b2127bbe7f8166869686630db2bf9b9ac1f41074d06fbaa5765addfa75ddaffd4c1854687c26ed960a1bff590f95ddc3a385a92b362f8d4ea4226b4531c8482a2f2432c062340b4f70d57f42db0733637d844d05c542572c8d9bed954a9afa244e704b4ac5d087b664a60b7b7bf316760c04b7a720f04803"   
}`

you can try altering the id or signature and it will give you an error message stating signature verification failure.

on successful runs, you should get a wallet verification failure

### CODE REVIEW

File: src/frontend/src/store/index.ts [hunk 1]
```diff
@@ -17,15 +21,44 @@ function uuidv4() {
...
 Vue.use(Vuex);
 Vue.use(VueToast);
 
 export default new Vuex.Store({
   state: {
     loading: false,
     walletCreated: false,
-    walletAmount: 0,
-    walletId: uuidv4(),
+    walletAmount: 50,
```
Comment (madisonfines) [inline src/frontend/src/store/index.ts]: Remember to change walletAmount back to 0 if changing to 50 was just for testing!

File: src/frontend/src/store/index.ts [hunk 2]
```diff
@@ -17,15 +21,44 @@ function uuidv4() {
...
+    walletId: publicKey,//uuidv4(),
+    privKey: privateKey,
```
Comment (madisonfines) [inline src/frontend/src/store/index.ts]: Nitpick: Could you change privKey to privateKey for clarity please

File: src/frontend/src/store/index.ts [hunk 3]
```diff
@@ -2,12 +2,16 @@ import axios from "axios";
 import Vue from "vue";
 import Vuex from "vuex";
 import VueToast from 'vue-toast-notification';
+import { sha256 } from "js-sha256";
 import 'vue-toast-notification/dist/theme-sugar.css';
+import * as rs from 'jsrsasign';
```
Comment (madisonfines) [inline src/frontend/src/store/index.ts:8]: Could you add two comments surrounding the creating signature section? So write "// CREATE TRANSACTION SIGNATURE" and then right before Vue.use(Vuex) write "// END". That way it's not just random code happening before the store section, and it has an explanation for being there.

File: src/frontend/src/store/index.ts [hunk 4]
```diff
@@ -57,6 +93,7 @@ export default new Vuex.Store({
       axios
         .post("http://localhost/wallet/create", {
           "walletId": getters.walletId,
+
```
Comment (madisonfines) [inline src/frontend/src/store/index.ts]: Don't need this extra line!

**Review by madisonfines — CHANGES_REQUESTED** [review]

File: src/frontend/src/store/index.ts (hunk 1 above)
Comment (abhisachdev17) [reply src/frontend/src/store/index.ts]: done

File: src/frontend/src/store/index.ts (hunk 2 above)
Comment (abhisachdev17) [reply src/frontend/src/store/index.ts]: done

File: src/frontend/src/store/index.ts (hunk 3 above)
Comment (abhisachdev17) [reply src/frontend/src/store/index.ts:8]: done!

File: src/frontend/src/store/index.ts (hunk 4 above)
Comment (abhisachdev17) [reply src/frontend/src/store/index.ts]: removed

File: .vscode/launch.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
+{
```
Comment (ZackHolmberg) [inline .vscode/launch.json]: This probably shouldn't be added

File: src/backend/services/transaction-api/src/routes.py [hunk 6]
```diff
@@ -16,6 +19,15 @@
...
 from shared import HttpCode, FailureReturnString
 
+def validateSignature(id, signature, address):
+    public_key = RSA.import_key(address)
+    unhexify = codecs.getdecoder('hex')
+    signature = unhexify(signature.encode("utf-8"))[0]
+
+    verifier = PKCS1_v1_5.new(public_key)
+
+    return(verifier.verify(SHA256.new(str.encode(id)), signature))
```
Comment (ZackHolmberg) [inline src/backend/services/transaction-api/src/routes.py]: Small nitpick for the sake of readability, can you put this into a variable and then return that variable instead please

File: src/backend/services/transaction-api/src/routes.py [hunk 7]
```diff
@@ -32,13 +44,17 @@ def createTransaction():
...
 
     except Exception as e:
         return jsonify(err=FailureReturnString.INCORRECT_PAYLOAD.value), HttpCode.BAD_REQUEST.value
 
     try:
-        isVerified = True
+        isVerified = validateSignature(transaction_id, signature, from_address)
         # TODO do verification if the user is logged in/ the other user exists etc
+        if not isVerified:
+            return jsonify(err="signature verification failed"), HttpCode.BAD_REQUEST.value
```
Comment (ZackHolmberg) [inline src/backend/services/transaction-api/src/routes.py]: This error string can be added into the shared file with the other error strings inside the corresponding ENUM

File: src/frontend/src/store/index.ts [hunk 8]
```diff
@@ -2,21 +2,46 @@ import axios from "axios";
...
 
-function uuidv4() {
-  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
-    const r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
-    return v.toString(16);
-  });
+// Transaction Signing
+const genKeyPair = (): string[] => {
+  const keyPair = rs.KEYUTIL.generateKeypair("RSA", 1024);
+  console.log(keyPair)
```
Comment (ZackHolmberg) [inline src/frontend/src/store/index.ts]: Leftover log

File: src/frontend/src/store/index.ts [hunk 9]
```diff
@@ -2,21 +2,46 @@ import axios from "axios";
...
+const sign = (transaction: Transaction, privateKey: string): string => {
+  const dataToSign = transaction.id;
+  const sig = new rs.KJUR.crypto.Signature({alg: 'SHA256withRSA'})
+
+  sig.init(privateKey);
+  sig.updateString(dataToSign);
+
+  return sig.sign();
+};
+// end
```
Comment (ZackHolmberg) [inline src/frontend/src/store/index.ts]: Can we please change this and the above comment to something more obivous like the below:

`// ================== END ==================`

File: src/frontend/src/store/index.ts [hunk 10]
```diff
@@ -78,18 +107,25 @@ export default new Vuex.Store({
...
           "from": transaction.from,
           "to": transaction.to,
           "amount": transaction.amount,
+          "id": transaction.id,
+          "signature": transaction.signature
         })
         .then((response) => {
           if(response.data.success) {
-            Vue.$toast.success('Transaction has sent!', {
+            Vue.$toast.success('Transaction has been sent!', {
```
Comment (ZackHolmberg) [inline src/frontend/src/store/index.ts]: Maybe

`Transaction sent successfully!` ?

**Review by ZackHolmberg — CHANGES_REQUESTED** [review]

**Review by ZackHolmberg — APPROVED** [review]
👍

**Review by madisonfines — APPROVED** [review]

**Pull request merged**